### DISCUSSION TITLE
Did Solar Power API Responses Change?

### DISCUSSION THREAD

**Discussion by GeorgeIoak** [timdorr/tesla-api#348] [Q&A] [answered]
I had code working for more than a month after the OAuth changes. I think my token expired earlier this week and I've been able to get new tokens but my code that was working isn't now. Here's what I send:

```https://owner-api.teslamotors.com/api/1/energy_sites/mysite_id/calendar_history?kind=power&end_date=2021-03-10T07:59:59Z```

where mysite_id is replaced my actual 12 digit ID and now when I check the response I'm seeing this:
```
timestamp  solar_power  battery_power  grid_power  grid_services_power  generator_power
0  2021-03-10T00:00:00-07:00            0              0           0                    0                0
1  2021-03-10T00:15:00-07:00            0              0           0                    0                0
2  2021-03-10T00:30:00-07:00            0              0           0                    0                0
3  2021-03-10T00:45:00-07:00            0              0           0                    0                0
```

Where this is what I used to get (96 readings for each day):
```
2020-10-31T08:00:00-08:00 1604160000 240
2020-10-31T08:15:00-08:00 1604160900 520
2020-10-31T08:30:00-08:00 1604161800 1080
2020-10-31T08:45:00-08:00 1604162700 1960
2020-10-31T09:00:00-08:00 1604163600 2920
2020-10-31T09:15:00-08:00 1604164500 3600
2020-10-31T09:30:00-08:00 1604165400 4120
```
No matter what I check I only get back 4 records and the data is always 0. I'm able to "get products" which returns the serial number which I believe tells me that I did successfully get a new token so I'm at a loss as to what's going on.

**Comment by briangru** [comment]
With a query like the one above, I'm now getting 5 minute data like what you originally had displayed above.
I haven't seen good documentation on the energy site API's.  (I knew the calendar_history endpoint existed but didn't know the parameters until you posted this.)  I'm interested in anything you can share about supported endpoints.

**Reply by GeorgeIoak** [reply]
I just tried my program again and I'm not getting the 15 minute data sets. Even on dates that I grabbed successfully before things changed on me.

Looking through these discussions I did find one new end point that is the only one that returns data for me and it's `live_status` so my URL looks like:
```
https://owner-api.teslamotors.com/api/1/energy_sites/mysite_id/live_status
```
and the response comes back as:
```
Result =  {'solar_power': 5440, 'load_power': 5440, 'grid_status': 'Unknown', 'grid_services_active': False, 'grid_power': 0, 'grid_services_power': 0, 'timestamp': '2021-03-21T09:57:43-07:00'}
```
I don't have a powerwall just Tesla Solar and the SolarEdge Inverter which wirelessly connects to the Tesla small black box gateway.

I wonder if things are changing again and I just happen to be one of the first to see the changes? The Android app seems to behave exactly like before and I can retrieve history on it. All the data is encrypted with TLS 1.3 so my Wireshark captures don't help me.

**Reply by GeorgeIoak** [reply]
Well I know I looked an API page and I didn't see the changes but looking this morning I found some changes on https://www.teslaapi.io/energy-sites/state-and-settings. I was using `calendar_history` but I see that has been changed to just `history`. This returns the 15 minute increment data like before but I haven't found the correct syntax to specify a start and end date that was working before.

**Reply by GeorgeIoak** [reply]
I figured I'd document some of my testing results in case it helps get to the bottom of this. I've experimented with the endpoints and found this:

using `kind:energy` instead of `kind:power` together with an endpoint of `history`
```
https://owner-api.teslamotors.com/api/1/energy_sites/my-site-id/history

{'kind': 'energy', 'period': 'day', 'end_date': '2021-03-10T07:59:59Z'}

{'response': {'serial_number': 'this-was-my-serial-number', 'period': 'day', 'time_series': [{'timestamp': '2021-03-20T01:00:00-07:00', 'solar_energy_exported': 58890, 'generator_energy_exported': 0, 'grid_energy_imported': 0, 'grid_services_energy_imported': 0, 'grid_services_energy_exported': 0, 'grid_energy_exported_from_solar': 0, 'grid_energy_exported_from_generator': 0, 'grid_energy_exported_from_battery': 0, 'battery_energy_exported': 0, 'battery_energy_imported_from_grid': 0, 'battery_energy_imported_from_solar': 0, 'battery_energy_imported_from_generator': 0, 'consumer_energy_imported_from_grid': 0, 'consumer_energy_imported_from_solar': 0, 'consumer_energy_imported_from_battery': 0, 'consumer_energy_imported_from_generator': 0}, {'timestamp': '2021-03-21T01:00:00-07:00', 'solar_energy_exported': 58900.000000000015, 'generator_energy_exported': 0, 'grid_energy_imported': 0, 'grid_services_energy_imported': 0, 'grid_services_energy_exported': 0, 'grid_energy_exported_from_solar': 0, 'grid_energy_exported_from_generator': 0, 'grid_energy_exported_from_battery': 0, 'battery_energy_exported': 0, 'battery_energy_imported_from_grid': 0, 'battery_energy_imported_from_solar': 0, 'battery_energy_imported_from_generator': 0, 'consumer_energy_imported_from_grid': 0, 'consumer_energy_imported_from_solar': 0, 'consumer_energy_imported_from_battery': 0, 'consumer_energy_imported_from_generator': 0}]}}
```
Note that even though the request date was for 2021-03-10 the data returned was for "today" and only 2 data sets were returned.

Now changing back to the original `calendar_history` and `kind:energy` does actually return data from 2021-03-10 but the data now has a time_zone_offset field and it's only 1 data point.

```
https://owner-api.teslamotors.com/api/1/energy_sites/my-site-id/calendar_history

{'kind': 'energy', 'period': 'day', 'end_date': '2021-03-10T07:59:59Z'}

{'response': {'serial_number': 'this-was-my-serial-number', 'period': 'day', 'time_zone_offset': -420, 'time_series': [{'timestamp': '2021-03-10T01:00:00-07:00', 'solar_energy_exported': 0, 'generator_energy_exported': 0, 'grid_energy_imported': 0, 'grid_services_energy_imported': 0, 'grid_services_energy_exported': 0, 'grid_energy_exported_from_solar': 0, 'grid_energy_exported_from_generator': 0, 'grid_energy_exported_from_battery': 0, 'battery_energy_exported': 0, 'battery_energy_imported_from_grid': 0, 'battery_energy_imported_from_solar': 0, 'battery_energy_imported_from_generator': 0, 'consumer_energy_imported_from_grid': 0, 'consumer_energy_imported_from_solar': 0, 'consumer_energy_imported_from_battery': 0, 'consumer_energy_imported_from_generator': 0}]}}
```
When using an endpoint of `history' and `kind:power` ignored the `end_date` parameter and returns all data from yesterday up to now in 15 minute increments.

So it seems like you can get historical data but now it defaults to only returning the 1st 4 data points so maybe it's expecting a "number of points" parameter or something to indicate that you want all 96 data points for `end_date`.

**Comment by GeorgeIoak** [comment, marked as answer]
So it seems that all I needed to do was to adjust the `end_date` hour. Before passing hour 7 resulted in all 96 data points but I guess because of daylight savings change I now have to pass hour 6 in order to get all the data. 

```
https://owner-api.teslamotors.com/api/1/energy_sites/my-site_id/calendar_history?kind=power&end_date=2021-03-20T06:59:59Z
```
I was curious how the hour impacted the returned data so I looped through all 24 hours and this was the result:

Hour| Max Returned Data |       # of Data Returned
----|-----------------------|--------------------------
0   | 2021-03-09T17:45:00-07:00    |  72
1   | 2021-03-09T18:45:00-07:00    |  76
2   | 2021-03-09T19:45:00-07:00    |  80
3   | 2021-03-09T20:45:00-07:00    |  84
4   | 2021-03-09T21:45:00-07:00    |  88
5   | 2021-03-09T22:45:00-07:00    |  92
6   | 2021-03-09T23:45:00-07:00    |  96
7   | 2021-03-10T00:45:00-07:00    |   4
8   | 2021-03-10T01:45:00-07:00    |   8
9   | 2021-03-10T02:45:00-07:00    |  12
10  | 2021-03-10T03:45:00-07:00    |  16
11  | 2021-03-10T04:45:00-07:00    |  20
12  | 2021-03-10T05:45:00-07:00    |  24
13  | 2021-03-10T06:45:00-07:00    |  28
14  | 2021-03-10T07:45:00-07:00    |  32
15  | 2021-03-10T08:45:00-07:00    |  36
16  | 2021-03-10T09:45:00-07:00    |  40
17  | 2021-03-10T10:45:00-07:00    |  44
18  | 2021-03-10T11:45:00-07:00    |  48
19  | 2021-03-10T12:45:00-07:00    |  52
20  | 2021-03-10T13:45:00-07:00    |  56
21  | 2021-03-10T14:45:00-07:00    |  60
22  | 2021-03-10T15:45:00-07:00    |  64
